nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/QuerySObject.java

```diff
@@ -0,0 +1,499 @@
+
+        if (validationContext.getProperty(AGE_FIELD).isSet() && !validationContext.getProperty(TIMESTAMP_FORMAT).isSet()) {
+            results.add(
+                new ValidationResult.Builder()
+                    .subject(AGE_FIELD.getDisplayName())
+                    .valid(false)
+                    .explanation("it requires " + TIMESTAMP_FORMAT.getDisplayName() + " also to be set.")
+                    .build()
+            );
+        }
+
+        if (validationContext.getProperty(INITIAL_AGE_FILTER).isSet() && !validationContext.getProperty(AGE_FIELD).isSet()) {
+            results.add(
+                new ValidationResult.Builder()
+                    .subject(INITIAL_AGE_FILTER.getDisplayName())
+                    .valid(false)
+                    .explanation("it requires " + AGE_FIELD.getDisplayName() + " also to be set.")
+                    .build()
+            );
+        }
+
+        return results;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        String sObject = context.getProperty(SOBJECT_NAME).getValue();
+        String fields = context.getProperty(FIELD_NAMES).getValue();
+        String customWhereClause = context.getProperty(CUSTOM_WHERE_CONDITION).getValue();
+        String dateFormat = context.getProperty(DateTimeUtils.DATE_FORMAT).getValue();
+        String timeFormat = context.getProperty(DateTimeUtils.TIME_FORMAT).getValue();
+        String timestampFormat = context.getProperty(DateTimeUtils.TIMESTAMP_FORMAT).getValue();
+        RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER).asControllerService(RecordSetWriterFactory.class);
+        boolean includeZeroRecordFlowFiles = context.getProperty(INCLUDE_ZERO_RECORD_FLOWFILES).asBoolean();
+
+        String ageField = context.getProperty(AGE_FIELD).getValue();
+        String initialAgeFilter = context.getProperty(INITIAL_AGE_FILTER).getValue();
+        Long ageDelayMs = context.getProperty(AGE_DELAY).asTimePeriod(TimeUnit.MILLISECONDS);
+
+        String ageFilterLower;
+        StateMap state;
+        try {
+            state = context.getStateManager().getState(Scope.CLUSTER);
+            ageFilterLower = state.get(LAST_AGE_FILTER);
+        } catch (IOException e) {
+            getLogger().error("Couldn't retrieve state", e);
+            throw new ProcessException(e);
+        }
+
+        String ageFilterUpper;
+        if (ageField == null) {
+            ageFilterUpper = null;
+        } else {
+            Instant ageFilterUpperTime;
+            if (ageDelayMs == null) {
+                ageFilterUpperTime = Instant.now();
+            } else {
+                ageFilterUpperTime = Instant.now().minus(ageDelayMs, ChronoUnit.MILLIS);
+            }
+
+            ageFilterUpper = new SimpleDateFormat(timestampFormat).format(java.sql.Timestamp.from(ageFilterUpperTime));
```

Comment:
Using DateTimeFormatter with the LocalDate pattern can result in the full form of localized offset concatenated to the formatted date string which salesforce cannot handle.

The `timestampFromat` property is derived from the `DateTimeUtils.timeStampFormat` property which specifies that the value must match the Java Simple Date Format. Since DateTimeFormatter uses different format patterns, the following default pattern `"yyyy-MM-dd'T'HH:mm:ss.SSSZZZZ"` would result in two different date strings.
- With LocalDateTime:         2022-03-11T00:34:40.210+0100
- With DateTimeFormatter: 2022-03-11T00:34:40.210GMT+01:00

From the DateTimeFormatter documentation: Offset Z: "Four letters outputs the full form of localized offset".

I can change the java.sql.TimeStamp.from to Date.from.